### ISSUE TITLE
Why does AttentionBlock perform softmax in float32?

### ISSUE THREAD

**Issue by cmdr2** [huggingface/diffusers#3483]
Hi,

// *This question also applies to the newly refactored replacement for AttentionBlock, which sets [`upcast_softmax=True`](https://github.com/huggingface/diffusers/blob/88d269461ca9b5acfae3dedd732438266f526109/src/diffusers/models/unet_2d_blocks.py#L1173) for the replacements of AttentionBlock.*

I'm curious to know why softmax is performed in float32 in AttentionBlock, instead of using whatever dtype the query is in (usually float16)? It doesn't look like CompVis/StabilityAI's original implementation does it in float32 either.

I don't know enough about the theory behind attention, so apologies if this is a newbie question. I tried comparing the images (with and without upcasting) for many models, and they looked the same. I also tried to trace the code ancestry (details below), and couldn't find this in the original Stable Diffusion code (it performs softmax in float16).

The upcast results in significantly higher peak VRAM consumption, so I thought I'd just ask :)

For e.g. in the erstwhile [AttentionBlock](https://github.com/huggingface/diffusers/blob/28f404349d69da1af7b52f18b022bc7971951a41/src/diffusers/models/attention.py#L181)
```py
attention_probs = torch.softmax(attention_scores.float(), dim=-1).type(attention_scores.dtype)
```

or in the new code that replaces AttentionBlock with Attention: [`upcast_softmax=True`](https://github.com/huggingface/diffusers/blob/88d269461ca9b5acfae3dedd732438266f526109/src/diffusers/models/unet_2d_blocks.py#L1173)

## Impact
For a 1024x1024 image (with attn slicing == 4), it results in an extra 6 GB of peak VRAM usage.

This increases the peak VRAM usage quite a bit (especially for large images), because a float32 tensor of shape `attention_scores` will be created first, and then an equal-sized `attention_probs` tensor will be created in float32, and finally downcast to float16.

If `attention_scores` is 2 GB with float16 (e.g. 1024x1024, with attention slicing == 4), this will consume 4 GB for the temp upcast `attention_scores`, and another 4 GB for the temp `attention_probs` (before it downcasts back to fp16). So **extra 8 GB of peak VRAM** usage.

If softmax wasn't upcast to float32, then we'd consume only an **extra 2 GB of VRAM** for making `attention_probs`.

## Does it produce different images?

I generated a few images using `SD 1.4`, `1.5`, `2.0-512`, `2.0-768-v` and `SD 2.1-768` (which requires `upcast_attn` to fp32), and performed the softmax in float16 and float32. For `PNDM` and `euler_a`.

I didn't see any visual difference in the images. A numpy difference shows a couple of pixels have opposite red or green or blue. Two example images below, using 2.1-768. I can share the other images if you'd like.

## Ancestry of the code

The earliest presence of this code in diffusers is [here](https://github.com/huggingface/diffusers/commit/fff981df2f5529118a8b5a8566a1cdc2d9910d63#diff-6c9e07391a079954ebec4daf2e6d8c29d2cb746d7cd3498c8a6a2d3553038016R138) from June 2022.

[StabilityAI/CompVis's implementation](https://github.com/Stability-AI/stablediffusion/blob/main/ldm/modules/diffusionmodules/model.py#L193) also doesn't use float32 for softmax. I printed the tensor info for what's sent into softmax, and it shows as float16.

The `upcast_softmax` flag was added to `Attention` for the [unCLIP PR](https://github.com/huggingface/diffusers/commit/2dcf64b72a44dc748610c3eb51534b325df3bca9), but I'm not sure whether this is necessary for regular SD pipelines as well?

@patrickvonplaten Do you remember if this upcast for softmax was added for better numerical stability? Is it required in every case?

---

Again, apologies if this is a newbie question. Thought it would be good to check, incase this was a low-hanging fruit.

Thanks!

float32 softmax (2.1-768): ![euler_a_0](https://github.com/huggingface/diffusers/assets/844287/b33efe38-c238-4b7c-84e7-15139312ae6c)
float16 softmax (2.1-768): ![euler_a_0](https://github.com/huggingface/diffusers/assets/844287/33892387-5508-41fa-8d35-fe7e8ccb0687)

**Comment by cmdr2** [comment]
Maybe it's okay to leave it at float32 for training, but allow a user to use float16 for inference?

**Comment by cmdr2** [comment]
FWIW, Easy Diffusion (beta) has been running diffusers with float16 softmax for about 4 days now, and no issues have been reported by users. I monkey-patched the diffusers pipeline to prevent upcasting to float32 before running the softmax.

The users on beta are quite happy with the lower VRAM usage (which allows them to create larger images, or batch more in parallel).

I'll update here if any issues are reported with softmax on float16.

Thanks!

**Comment by patrickvonplaten** [comment]
Hey @cmdr2,

Thanks for opening this detailed issue. We have added "force-float32" to `AttentionBlock` back in the days to make sure we fully match CompVis' orig code here: https://github.com/CompVis/stable-diffusion/blob/21f890f9da3cfbeaba8e2ac3c425ee9e998d5229/ldm/modules/diffusionmodules/openaimodel.py#L404

But it indeed seems that this upscaling is now not needed anymore. This is also why we've deleted/fully deprecated this class.

**Comment by cmdr2** [comment]
Hi @patrickvonplaten Thanks for the clarification, that makes sense!

So is `upcast_softmax=True` necessary in the places where the AttentionBlock class was replaced with the unified Attention class? E.g. https://github.com/huggingface/diffusers/blob/64bf5d33b7ef1b1deac256bed7bd99b55020c4e0/src/diffusers/models/unet_2d_blocks.py#L1134

`upcast_softmax=True` was added in the recent "attention refactor" PR, in all the places where AttentionBlock was replaced - https://github.com/huggingface/diffusers/commit/909742dbd6873052995dc6cd5f4150ff238015d2#diff-6772eedf129163896c86d8423027ba04081212c2733ce1eca3ce98f9208b7670R1082

If it's not necessary, I'd be happy to submit a PR.

Two options:
1. Continue this behavior by default, and provide a user-facing function to disable `upcast_softmax` by looping through all the attention blocks. But while this feels safer, it continues a high VRAM-usage pattern for default users, especially if float32 softmax is not really necessary.

2. If we're sure our unit/integration tests are strong enough to catch regressions, we can just set `upcast_softmax=False` in all the places (where AttentionBlock was previously) and cross our fingers? And reap the benefits of lower VRAM usage for everyone (larger image sizes and bigger batch sizes).

Thanks again for looking into this! I appreciate it.

**Comment by cmdr2** [comment]
I'll dig around a bit more to confirm if float32 softmax is still being used in regular SD.

**Comment by cmdr2** [comment]
Thanks! That makes sense. I'll dig into this a bit more, and verify whether float32 softmax is still required, and used in regular SD.

The fact that it's running fine (5 days on Easy Diffusion beta) with float16 softmax seems interesting, and the VRAM savings are appealing. I'll research this a bit more and post here.

**Comment by cmdr2** [comment]
Will provide an update in about 2 days. Looks okay so far.

**Comment by cmdr2** [comment]
Hi, sorry for the delay. Here are my findings and recommendation.

tl;dr - float32 softmax is important for training, but probably not necessary for inference. It's been running fine on Easy Diffusion beta for over 5 weeks now. I recommend keeping the default as float32 (i.e. no change), but expose a function to let a user switch to float16 softmax, like a memory optimization utility function.

**Details:**

float16 softmax() has been running for about 5 weeks in Easy Diffusion beta now, and I haven't heard any complaints. A lot of Easy Diffusion's power users use the beta version, and are quite vocal whenever something breaks. So in practice, this change seems to work well.

I dug into the code again, and it seems like softmax is still performed in float32 even in Stability AI's official implementation. https://github.com/Stability-AI/stablediffusion/blob/main/ldm/modules/diffusionmodules/openaimodel.py#L403

So I was wrong about it not being present in Stability AI's official implementation - they do perform softmax in float32, even today.

There's even a change (5 years ago) to change softmax to use float32, so it's clearly something people prefer - https://github.com/facebookresearch/fairseq/commit/d6be0c7e001c51fc7639304a463dae4041d535f6

From what I can understand, float32 softmax is important during training. For e.g. the recommendation here (for training) is clear that softmax should be performed in float32 (bullet point 4) - https://docs.nvidia.com/deeplearning/performance/mixed-precision-training/index.html#training_tensorflow

But that advice is for training, not inference. Based on empirical evidence, maybe it's okay to use float16 for inference? As an optimization option?

**Recommendation:**

So my recommendation is to continue using float32 softmax as the default behavior, but provide a user-facing function to switch the model to use float16 softmax for inference. This function can be listed in the "Memory and Optimizations" section.

@patrickvonplaten @sayakpaul What do you feel about this change? Thanks!

**Comment by cmdr2** [comment]
I'm also happy to leave things as they are in diffusers, if you'd prefer. I can write a utility function in Easy Diffusion's code to loop through the attention modules and set `upcast_softmax` to `False`. So we aren't blocked on this change.

The impact of this change is significant for larger images. E.g. for a 1024x1024 image (with attn slicing == 4), it reduces the peak VRAM usage by 6 GB.

**Comment by sayakpaul** [comment]
Thanks for so much for being generous with your details. To me exposing an argument to not upcast the softmax during inference sounds reasonable given the memory savings. 

WDYT? @patrickvonplaten @williamberman

**Comment by patrickvonplaten** [comment]
I think this issue is moot because we don't use this attention block: https://github.com/huggingface/diffusers/blob/28f404349d69da1af7b52f18b022bc7971951a41/src/diffusers/models/attention.py#L50 anywhere anymore

**Comment by williamberman** [comment]
There is still an upcast softmax in the default AttentionProcessor, however I'm against exposing an argument to turn it off. The recommended way to get memory savings should be to use the flash attention kernels

**Comment by cmdr2** [comment]
@patrickvonplaten The refactoring of attention doesn't change the overall process. The float32 softmax is just performed in a different class. It's now performed here: https://github.com/huggingface/diffusers/blob/88d269461ca9b5acfae3dedd732438266f526109/src/diffusers/models/unet_2d_blocks.py#L1173

> The recommended way to get memory savings should be to use the flash attention kernels

I can see the logic with that. If the recommendation is that people who want to further optimize the memory usage should either use xformers or pytorch 2.0 (with [scale_qk](https://github.com/huggingface/diffusers/blob/88d269461ca9b5acfae3dedd732438266f526109/src/diffusers/models/attention_processor.py#L161)), I'm fine with closing this issue.

The purpose of this issue was to investigate potentially unnecessary VRAM usage in the stock diffusers implementation (upto 6 GB extra for a 1024x1024 image). I can understand not wanting to touch the logic too much.

I suspect the float32 softmax (for inference) is an artifact (someone probably did that once, and that's propagated in every implementation since).

Either case, I'm fine with closing this issue.

**Comment by patrickvonplaten** [comment]
Hey @cmdr2, 

Note that we don't use this attention block anymore

**Comment by cmdr2** [comment]
@patrickvonplaten It's not a question about this particular attention block. Like william mentioned and I've mentioned (in my previous comment, and in the OP), the replacement for AttentionBlock also upcasts the softmax before performing it. https://github.com/huggingface/diffusers/blob/88d269461ca9b5acfae3dedd732438266f526109/src/diffusers/models/unet_2d_blocks.py#L1173

**Comment by cmdr2** [comment]
Closing this, I'm an idiot :) Not sure how the new implementation avoids the float32 softmax (I swear I saw that code), but I profiled the memory usage with diffusers 0.18.1, and it's really good. Nice work!
